### ISSUE TITLE
rules priority clarification?

### ISSUE THREAD

**Issue by aupuzikov** [espeak-ng/espeak-ng#1679]
Hi guys! Can someone explain rules priority applying . I have a couple of examples
### for example ### 
for spanish rules from [here](https://github.com/espeak-ng/espeak-ng/blob/master/dictsource/es_rules) 
```
apuzikov@gpu-test:~$ espeak-ng "graduó" -Xq -ves-la --ipa=2
Unpronouncable? 'graduó'
 39	_) g (L01A [g]
  5	_) g ($unpr  [g]
Translate 'graduó'
  1	g        [g]
 39	_) g (L01A [g]
  5	_) g (   [g]
 21	g (L03   [Q]

  1	r        [r]
 40	C) r (A  [r]

  1	a        [a]

  1	d        [d]
 21	d (A     [D]

  1	u        [u]
 21	u (A     [w]

 22	ó       [''o]

 ɡɾaðwˈo
```

as we see for the word 'graduó'. For it's first grapheme we have plenty of matching rules. as I see from description rule #21 is latest so it's the one which must be applied. but this isn't what we see here.

### Another example:
```apuzikov@gpu-test:~$ espeak-ng "hablando" -Xq -ves --ipa=0
Translate 'hablando'
  1	h        []

  1	a        [a]

  1	b        [b]

  1	l        [l]

  1	a        [a]

  1	n        [n]

  1	d        [d]
 22	n) d     [d]
 21	d (A     [D]

  1	o        [o]

 aβlˈando
```

for the second grapheme there is no rule for b to become B(β), still it becomes a B(β). Why ?

Version is 
```apuzikov@gpu-test:~$ espeak-ng --version
eSpeak NG text-to-speech: 1.49.2  Data at: /usr/lib/x86_64-linux-gnu/espeak-ng-data
```

**Comment by jaacoppi** [comment]
> as we see for the word 'graduó' and it's first grapheme we have plenty of matching rules. as I see from description 21 rule is latest so it's the one which must be applied. but this isn't what we see here.

Highest score gets applied.

> for the second grapheme there is no rule for b to become B(β), still it becomes a B(β). Why ?

Interesting! Looks like a bug.

**Comment by aupuzikov** [comment]
> Highest score gets applied.
Juho, hi! thank you for quick response! yeah I see there are **points**  in **MatchRecord** struct. Is there any handwritten documentation of how to evaluate those points or there is only raw code?

**Comment by jaacoppi** [comment]
There's something in docs/dictionary.md but unfortunately the details are in the code only. Basically a more detailed rule match gets more points.

**Comment by aupuzikov** [comment]
> Interesting! Looks like a bug.

can we determine the cause? and maybe fix it?

looks like it is a common issue. For example ***d*** grapheme
```
apuzikov@gpu-test:~$ espeak-ng "jardín" -Xq -ves --ipa
Translate 'jardín'
  1	j        [x]

  1	a        [a]

  1	r        [r]

  1	d        [d]
 22	r) d     [d]
 21	d (A     [D]

 22	í       [''i]

  1	n        [n]

 xaɾðˈin
```

**Comment by jaacoppi** [comment]
Found it. The reason is the keyword ChangePhoneme in phsource/ph_spanish:

phoneme d
  vcd dnt stp

  IF nextPh(isVowel) OR nextPh(isRhotic) OR nextPh(j) THEN
    IF prevPh(isPause) OR prevPh(isNasal) OR prevPh(l) THEN
    ELSE
      ChangePhoneme(D)
    ENDIF
  ENDIF

Would be useful to add a debug print when a phonemechange happens.

**Comment by aupuzikov** [comment]
great! Thank you for clarification. Are these rules applied strictly after the ones in "es_rules"? can you point out the place in source code where these rules are used?

**Comment by jaacoppi** [comment]
Searching for "ChangePhoneme" shows that src/libespeak-ng/compiledata.c has a struct with all the phoneme keywords. Searching where that struct is used leads to:
src/libespeak-ng/compiledata.c:1857:			case i_CHANGE_PHONEME

I don't remember what compiledata does or how it fits the program flow.

**Comment by aupuzikov** [comment]
ok, but there is still a problem with **hablando**
```
Translate 'hablando'
  1	h        []

  1	a        [a]

  1	b        [b]

  1	l        [l]

  1	a        [a]

  1	n        [n]

  1	d        [d]
 22	n) d     [d]
 21	d (A     [D]

  1	o        [o]

 aβlˈando

```
there is definitely a rule for letter **b**, here it is 
```
phoneme b
  vcd blb stp

  IF nextPh(isVowel) OR nextPh(isRhotic) OR nextPh(isLiquid) THEN
    IF prevPh(isPause) OR prevPh(isNasal) THEN
    ELSE
      ChangePhoneme(B)
    ENDIF
  ENDIF

  CALL base1/b
endphoneme
```
but it can't be applied, cause previous phoneme isn't a pause or a nasal, why there is β?

**Comment by aupuzikov** [comment]
sorry for spamming this issue with questions, but I don't quite sure still understand why is some rules applied
for ex
```
apuzikov@gpu-test:~$ espeak-ng "muy difícil" -Xq -ves-419 --ipa=2
Translate 'muy'
  1	m        [m]

  1	u        [u]

  1	y        [i]
 22	y (_     [i]
 42	u) y (K  [j:]

Translate 'difícil'
  1	d        [d]
 25	_) d (A  [d]
  5	_) d (   [d]
 21	d (A     [D]

  1	i        [i]

  1	f        [f]

 22	í       [''i]

  1	c        [k]
 21	c (Y     [T]

  1	i        [i]

  1	l        [l]

 mˈujj ðifˈisil
```
there is no **y** rule in [y-rules](https://github.com/espeak-ng/espeak-ng/blob/master/dictsource/es_rules#L254-L263) that corresponds to **"u) y (K  [j:]"** , where it came from?
Moreover, if it there still this rule, how can it be applied if there is 'space' between **y** and **d**. In the dictionary.md documentation there is description of <post> _"characters in (<post> group will not be spelled and produced, but can be used as reference to choose matching rule"_ . My understanding was that i should be like **"u) y (_K  [j:]"** to be applied here. Or is the 'space' ignored?

**Comment by aupuzikov** [comment]
```
apuzikov@gpu-test:~$ espeak-ng "rebelde" -Xq -ves-419 --ipa=2
Translate 'rebelde'
  1	r        [r]
 25	_) r (A  [rr]

  1	e        [e]

  1	b        [b]
 21	b (A     [B]

  1	e        [e]

  1	l        [l]

  1	d        [d]
 22	l) d     [d]
 21	d (A     [D]

  1	e        [e]

 ɾɾeβˈelðe
```
anyway this last **d** grapheme still not working out. There is [a rule](https://github.com/espeak-ng/espeak-ng/blob/1.49.2/phsource/ph_spanish#L111) for phoneme **d**. Here it is 
```
phoneme d
  vcd dnt stp

  IF nextPh(isVowel) OR nextPh(isRhotic) OR nextPh(j) THEN
    IF prevPh(isPause) OR prevPh(isNasal) OR prevPh(l) THEN
    ELSE
      ChangePhoneme(D)
    ENDIF
  ENDIF

  CALL base2/d
endphoneme
```
however this rule can't be applied cause in **rebeldo** previous phoneme for **d** is **l**, why there is **ð**?

**Comment by jaacoppi** [comment]
> ***@***.***:~$ espeak-ng "muy difícil" -Xq -ves-419 --ipa=2

Most likely you're looking at the current source but using an older version of espeak-ng.
That rule was changed in commit 748ae2e.

**Comment by jaacoppi** [comment]
> ***@***.***:~$ espeak-ng "rebelde" -Xq -ves-419 --ipa=2

You're mixing IPA and Kirshenbaum.

Kirshenbaum = D
IPA = ð

If you need to debug the rules you must omit the flag --ipa=2 OR be extremely fluent in both transcripts and be able to convert them in your head.

**Comment by aupuzikov** [comment]
> You're mixing IPA and Kirshenbaum.

yep sorry, let's stick with Kirshenbaum. the point is that there is no rule or **ChangePhoneme** call that changes **d** to **D** in example below
```
apuzikov@gpu-test:~$ espeak-ng "rebelde" -Xq -ves-419
Translate 'rebelde'
  1	r        [r]
 25	_) r (A  [rr]

  1	e        [e]

  1	b        [b]
 21	b (A     [B]

  1	e        [e]

  1	l        [l]

  1	d        [d]
 22	l) d     [d]
 21	d (A     [D]

  1	e        [e]

 r**eB'elDe
```

**Comment by jaacoppi** [comment]
Removing the IF...ENDIF for ChangePhoneme causes r**eB'elDe to become r**eB'elde. This means the ChangePhoneme takes effect even though it shouldn't.

My guess is the syntax for the phonemes in ph_spanish is wrong and the syntax parser of espeak-ng has a bug where ELSE is wrongly executed.

Most if statements in programming are like this:
IF (condition) 
	THEN (do something)
	ELSE (do something else)
ENDIF

That's how other phonemes like ph_english handle it. In ph_spanish the THEN part is missing.

Assuming I'm correct:
Quick solution: Write a correct IF statement for ph_spanish (requires language knowledge)
Good solution: First fix the bug in the parser, then implement the Quick solution.

**Comment by jaacoppi** [comment]
@aupuzikov Do you think this can be closed now?

**Comment by aupuzikov** [comment]
Juho, hi! I guess we can close it, but maybe someday somebody will make PR and fix these rules? =) Won't this issue be helpfull later?

**Comment by sevapopov2** [comment]
Hello everyone! That is great that there are going to be some fixes for Spanish in the future! As a person who learns Spanish it is really important for me. Do I understand correctly that it can fix the letter b pronounced in words as v and d as english th? That is the most frustrating thing for me why I don't prefer eSpeak for Spanish
